File: backend/internal/api/videos.go
package api

import (
	"errors"
	"log/slog"
	"net/http"
	"strconv"
	"strings"

	"homereel/backend/internal/domain"
	"homereel/backend/internal/scanner"
	"homereel/backend/internal/streaming"
)

// videoOrError loads a video, writing an error response when missing.
func (s *Server) videoOrError(w http.ResponseWriter, r *http.Request, id string) (*domain.Video, bool) {
	v, err := s.videos.Get(r.Context(), id)
	if err != nil {
		if errors.Is(err, domain.ErrNotFound) {
			writeError(w, http.StatusNotFound, "not_found", "视频不存在")
			return nil, false
		}
		slog.Error("get video", "err", err)
		writeError(w, http.StatusInternalServerError, "internal", "服务器内部错误")
		return nil, false
	}
	return &v, true
}

func (s *Server) handleVideosList(w http.ResponseWriter, r *http.Request) {
	page, _ := strconv.Atoi(r.URL.Query().Get("page"))
	pageSize, _ := strconv.Atoi(r.URL.Query().Get("pageSize"))
	res, err := s.videos.List(r.Context(), domain.VideoQuery{
		Q:         r.URL.Query().Get("q"),
		Kind:      r.URL.Query().Get("kind"),
		Tags:      r.URL.Query()["tag"],
		ShowID:    r.URL.Query().Get("showId"),
		Ungrouped: r.URL.Query().Get("ungrouped") == "1",
		Sort:      r.URL.Query().Get("sort"),
		Order:     r.URL.Query().Get("order"),
		Page:      page,
		PageSize:  pageSize,
	})
	if err != nil {
		slog.Error("list videos", "err", err)
		writeError(w, http.StatusInternalServerError, "internal", "服务器内部错误")
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{
		"videos":    res.Videos,
		"total":     res.Total,
		"page":      max(page, 1),
		"page_size": max(pageSize, 24),
	})
}

func (s *Server) handleVideoDetail(w http.ResponseWriter, r *http.Request) {
	v, ok := s.videoOrError(w, r, r.PathValue("id"))
	if !ok {
		return
	}
	tags, err := s.videos.Tags(r.Context(), v.ID)
	if err != nil {
		slog.Error("get video tags", "video_id", v.ID, "err", err)
		writeError(w, http.StatusInternalServerError, "internal", "服务器内部错误")
		return
	}
	seriesID := s.videoSeriesID(r.Context(), v)
	// 单集详情按需检查源文件：路径在 → ok；路径不在但能在媒体源内按 file_id
	// 找到（改名/移动）→ moved（可同步）；找不到 → missing（可移除）。
	status, checkErr := s.scanner.CheckVideo(r.Context(), v.ID)
	sourceStatus := "ok"
	if checkErr == nil {
		sourceStatus = status.Status
	} else {
		slog.Warn("check video source", "video_id", v.ID, "err", checkErr)
	}
	writeJSON(w, http.StatusOK, map[string]any{
		"video":              v,
		"tags":               tags,
		"series_id":          seriesID,
		"direct_playable":    s.streaming.DirectPlayable(*v),
		"remux_playable":     s.streaming.RemuxPlayable(*v),
		"transcode_playable": s.streaming.TranscodePlayable(*v),
		"source_status":      sourceStatus,
		"new_path":           status.Path,
	})
}

// handleVideoSync re-locates a video's source file by file_id in its media
// source (rename/move) and converges its series membership. A file that cannot
// be found returns 404 so the UI can offer removing the record.
func (s *Server) handleVideoSync(w http.ResponseWriter, r *http.Request) {
	id := r.PathValue("id")
	if _, ok := s.videoOrError(w, r, id); !ok {
		return
	}
	if err := s.scanner.SyncVideo(r.Context(), id); err != nil {
		if errors.Is(err, scanner.ErrVideoMissing) {
			writeError(w, http.StatusNotFound, "not_found", "源文件不存在，请移除该单集")
			return
		}
		slog.Error("sync video", "video_id", id, "err", err)
		writeError(w, http.StatusInternalServerError, "internal", "同步失败")
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"synced": true})
}

// streamOrError resolves the video for any source-dependent streaming work.
func (s *Server) streamOrError(w http.ResponseWriter, r *http.Request) (*domain.Video, bool) {
	return s.videoOrError(w, r, r.PathValue("id"))
}

func (s *Server) handleStreamDirect(w http.ResponseWriter, r *http.Request) {
	v, ok := s.streamOrError(w, r)
	if !ok {
		return
	}
	if err := s.streaming.Direct(w, r, *v); err != nil {
		if errors.Is(err, streaming.ErrUnavailable) {
			writeError(w, http.StatusConflict, "storage_unavailable", "源文件不可用（存储离线）")
			return
		}
		slog.Error("stream direct", "video_id", v.ID, "err", err)
		writeError(w, http.StatusInternalServerError, "internal", "服务器内部错误")
	}
}

// handleStreamRemux serves the cached remuxed MP4 of a video (ADR-006 修订):
// the frontend requests it when canPlayType rejected the original container but
// the streams are browser-decodable, so a stream copy makes the file playable
// natively over HTTP Range with full seek.
func (s *Server) handleStreamRemux(w http.ResponseWriter, r *http.Request) {
	v, ok := s.streamOrError(w, r)
	if !ok {
		return
	}
	if err := s.streaming.Remux(w, r, *v, audioIndex(r)); err != nil {
		if errors.Is(err, streaming.ErrUnavailable) {
			writeError(w, http.StatusConflict, "storage_unavailable", "源文件不可用（存储离线）")
			return
		}
		slog.Error("stream remux", "video_id", v.ID, "err", err)
		writeError(w, http.StatusInternalServerError, "internal", "转换失败")
	}
}

// handleStreamHLSPlaylist serves the VOD playlist of the HLS transcode stream for
// a video (ADR-006 修订): the frontend requests it when canPlayType rejected
// direct play and the video is not remuxable (codec incompatible). The playlist
// declares the full video timeline (segments snapped to keyframes), so hls.js
// renders a draggable progress bar instead of a live window. Segments are
// generated on demand at seg-{n}.ts.
func (s *Server) handleStreamHLSPlaylist(w http.ResponseWriter, r *http.Request) {
	v, ok := s.streamOrError(w, r)
	if !ok {
		return
	}
	if err := s.streaming.Playlist(w, r, *v, r.URL.Query().Get("session"), audioIndex(r)); err != nil {
		slog.Warn("hls playlist", "video_id", v.ID, "err", err)
		writeError(w, http.StatusConflict, "stream_unavailable", "该视频暂时无法动态流式播放")
	}
}

// audioIndex parses the "?audio=" track selector (default 0). A negative or
// unparsable value falls back to the default track.
func audioIndex(r *http.Request) int {
	n, err := strconv.Atoi(r.URL.Query().Get("audio"))
	if err != nil || n < 0 {
		return 0
	}
	return n
}

// handleStreamHLSSegment serves one on-demand HLS segment (seg-{n}.ts). Each
// request maps to exactly one segment of the VOD playlist; the backend generates
// it with ffmpeg when it is not already cached.
func (s *Server) handleStreamHLSSegment(w http.ResponseWriter, r *http.Request) {
	v, ok := s.streamOrError(w, r)
	if !ok {
		return
	}
	file := r.PathValue("file")
	n, ok := hlsSegmentIndex(file)
	if !ok {
		writeError(w, http.StatusNotFound, "not_found", "分片不存在")
		return
	}
	if err := s.streaming.Segment(w, r, *v, r.URL.Query().Get("session"), n); err != nil {
		if errors.Is(err, streaming.ErrNotFound) {
			writeError(w, http.StatusNotFound, "not_found", "分片不存在")
			return
		}
		slog.Warn("hls segment", "video_id", v.ID, "n", n, "err", err)
		writeError(w, http.StatusInternalServerError, "internal", "分片生成失败")
	}
}

// hlsSegmentIndex parses a "seg-N.ts" segment file name into its index.
func hlsSegmentIndex(file string) (int, bool) {
	if rest, ok := strings.CutPrefix(file, "seg-"); ok {
		if name, ok := strings.CutSuffix(rest, ".ts"); ok {
			if n, err := strconv.Atoi(name); err == nil && n >= 0 {
				return n, true
			}
		}
	}
	return 0, false
}

func (s *Server) handleStreamCover(w http.ResponseWriter, r *http.Request) {
	v, ok := s.videoOrError(w, r, r.PathValue("id"))
	if !ok {
		return
	}
	thumb := r.URL.Query().Get("thumb") == "1"
	if err := s.streaming.Cover(w, r, *v, thumb); err != nil {
		if errors.Is(err, streaming.ErrNotFound) {
			writeError(w, http.StatusNotFound, "not_found", "封面不存在")
			return
		}
		slog.Error("stream cover", "video_id", v.ID, "err", err)
		writeError(w, http.StatusInternalServerError, "internal", "服务器内部错误")
	}
}

func (s *Server) handleStreamSubtitle(w http.ResponseWriter, r *http.Request) {
	v, ok := s.streamOrError(w, r)
	if !ok {
		return
	}
	track := -1
	if t, err := strconv.Atoi(r.URL.Query().Get("track")); err == nil {
		track = t
	}
	if err := s.streaming.Subtitle(w, r, *v, track); err != nil {
		if errors.Is(err, streaming.ErrNotFound) {
			writeError(w, http.StatusNotFound, "not_found", "未找到字幕")
			return
		}
		slog.Error("stream subtitle", "video_id", v.ID, "err", err)
		writeError(w, http.StatusInternalServerError, "internal", "服务器内部错误")
	}
}

// handleVideoSubtitles returns every subtitle source of a video so the player
// can offer a track menu (sidecar file + embedded text tracks).
func (s *Server) handleVideoSubtitles(w http.ResponseWriter, r *http.Request) {
	v, ok := s.videoOrError(w, r, r.PathValue("id"))
	if !ok {
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"subtitles": s.streaming.ListSubtitles(r.Context(), *v)})
}

// handleVideoAudioTracks returns every audio track of a video so the player can
// offer an audio track menu (multi-track containers like 国语/粤语 MKVs).
func (s *Server) handleVideoAudioTracks(w http.ResponseWriter, r *http.Request) {
	v, ok := s.videoOrError(w, r, r.PathValue("id"))
	if !ok {
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"audio": s.streaming.ListAudioTracks(r.Context(), *v)})
}

File: frontend/src/api/videos.ts
import { api } from './client'

export interface Video {
  id: string
  source_id?: string
  file_id: string
  relative_path: string
  path: string
  size: number
  mtime: number
  title: string
  kind: 'movie' | 'episode'
  duration: number
  codec: string
  audio_codec?: string
  container: string
  segmented?: boolean
  faststart?: boolean
  width: number
  height: number
  fps?: number
  file_size?: number
  cover_path: string
  thumb_path: string
  backdrop_path?: string
  show_id?: string
  season_number?: number
  episode_number?: number
  episode_title?: string
  year?: number
  rating?: number
  genre?: string
  studio?: string
  cast_text?: string
  metadata_source: string
  created_at: string
  updated_at: string
  last_scanned_at: string
}

export interface VideoListResult {
  videos: Video[]
  total: number
  page: number
  page_size: number
}

export interface VideoQuery {
  q?: string
  genre?: string
  year?: number
  kind?: 'movie' | 'episode'
  tags?: string[]
  showId?: string
  ungrouped?: boolean
  sort?: 'title' | 'date' | 'duration' | 'name' | 'rating'
  order?: 'asc' | 'desc'
  page?: number
  pageSize?: number
}

// VideoDetailResponse is the shape of GET /api/videos/:id: the video plus the
// backend-computed playability flags and the on-demand source-file status
// (ok | moved | missing, 单集详情同步提示用).
//
// direct_playable is the conservative Range fallback; remux_playable gates the
// container-only MP4 remux (/api/stream/:id/remux, served over Range) and
// transcode_playable gates the on-demand HLS re-encode (/api/stream/:id/hls).
export interface VideoDetailResponse {
  video: Video
  tags: string[]
  series_id?: string
  direct_playable: boolean
  remux_playable: boolean
  transcode_playable: boolean
  source_status: 'ok' | 'moved' | 'missing'
  new_path?: string
}

export interface HistoryEntry {
  video_id: string
  user: string
  progress: number
  updated_at: string
}

export interface VideoPatch {
  title?: string
  kind?: 'movie' | 'episode'
  year?: number
  rating?: number
  genre?: string
  studio?: string
  cast_text?: string
  tags?: string[]
}

export interface TagCount {
  tag: string
  count: number
}

export function fetchVideos(query: VideoQuery = {}): Promise<VideoListResult> {
  const params = new URLSearchParams()
  if (query.q) params.set('q', query.q)
  if (query.genre) params.set('genre', query.genre)
  if (query.year) params.set('year', String(query.year))
  if (query.kind) params.set('kind', query.kind)
  for (const tag of query.tags ?? []) params.append('tag', tag)
  if (query.showId) params.set('showId', query.showId)
  if (query.ungrouped) params.set('ungrouped', '1')
  if (query.sort) params.set('sort', query.sort)
  if (query.order) params.set('order', query.order)
  if (query.page && query.page > 1) params.set('page', String(query.page))
  if (query.pageSize) params.set('pageSize', String(query.pageSize))
  const qs = params.toString()
  return api<VideoListResult>(`/api/videos${qs ? `?${qs}` : ''}`)
}

export function fetchVideo(id: string): Promise<VideoDetailResponse> {
  return api<VideoDetailResponse>(`/api/videos/${id}`)
}

export function updateVideo(id: string, patch: VideoPatch): Promise<{ video: Video }> {
  return api<{ video: Video }>(`/api/videos/${id}`, {
    method: 'PATCH',
    body: JSON.stringify(patch),
  })
}

export function deleteVideo(id: string): Promise<{ deleted: boolean }> {
  return api<{ deleted: boolean }>(`/api/videos/${id}`, { method: 'DELETE' })
}

export function syncVideo(id: string): Promise<{ synced: boolean }> {
  return api<{ synced: boolean }>(`/api/videos/${id}/sync`, { method: 'POST' })
}

export function fetchHistory(id: string): Promise<{ history: HistoryEntry | null }> {
  return api<{ history: HistoryEntry | null }>(`/api/videos/${id}/history`)
}

export function putHistory(id: string, progress: number): Promise<{ saved: boolean }> {
  return api<{ saved: boolean }>(`/api/videos/${id}/history`, {
    method: 'PUT',
    body: JSON.stringify({ progress }),
  })
}

export function clearHistory(id: string): Promise<{ cleared: boolean }> {
  return api<{ cleared: boolean }>(`/api/videos/${id}/history`, { method: 'DELETE' })
}

// PlaybackPrefs is the effective playback selection cache (ADR-006 player
// prefs): the chosen audio track, the chosen subtitle source and the volume.
// scope tells which record the values come from — "series" means the video's
// series shares one memory across every episode (tracks are matched by NAME
// against the current episode's own tracks), "video" the member's own concrete
// selections. The player auto-applies them on every play and only refreshes them
// when the user manually changes a selection. As a rebuildable cache it can be
// cleared from the cache manager page or the detail pages; deleting a row resets
// the video to the browser defaults.
export interface PlaybackPrefs {
  scope: 'series' | 'video'
  audio_track?: number
  subtitle_id?: string
  audio_track_name?: string
  subtitle_name?: string
  volume?: number
  muted?: boolean
  updated_at?: string
}

// VideoPrefsResponse carries the effective prefs plus the video's series context
// (series_id is present whenever the video belongs to a series, so the player
// writes series-scoped on manual changes even before a series record exists).
export interface VideoPrefsResponse {
  prefs: PlaybackPrefs | null
  series_id?: string
}

// PlaybackPrefsPatch is a partial update: only the provided fields are written.
// For a series member the player sends track NAMES (audio_track_name/subtitle_
// name) and volume/muted; subtitle_name is '' when subtitles were turned off.
// Standalone videos send the concrete per-video audio_track/subtitle_id.
export interface PlaybackPrefsPatch {
  audio_track?: number
  subtitle_id?: string
  audio_track_name?: string
  subtitle_name?: string
  volume?: number
  muted?: boolean
}

export function fetchVideoPrefs(id: string): Promise<VideoPrefsResponse> {
  return api<VideoPrefsResponse>(`/api/videos/${id}/prefs`)
}

export function putVideoPrefs(id: string, patch: PlaybackPrefsPatch): Promise<{ saved: boolean }> {
  return api<{ saved: boolean }>(`/api/videos/${id}/prefs`, {
    method: 'PUT',
    body: JSON.stringify(patch),
  })
}

export function fetchTags(): Promise<{ tags: TagCount[] }> {
  return api<{ tags: TagCount[] }>('/api/tags')
}

export function searchVideos(q: string): Promise<{ videos: Video[] }> {
  return api<{ videos: Video[] }>(`/api/search?q=${encodeURIComponent(q)}`)
}

export function streamUrl(id: string): string {
  return `/api/stream/${id}`
}

// remuxUrl returns the cached remuxed-MP4 stream URL (ADR-006 修订): served over
// HTTP Range like a native file, generated on first play for browser-decodable
// streams in a foreign container (e.g. MKV h264+aac). audio selects which audio
// track is mapped (default 0).
export function remuxUrl(id: string, audio = 0): string {
  return `/api/stream/${id}/remux${audio > 0 ? `?audio=${audio}` : ''}`
}

// hlsPlaylistUrl returns the VOD playlist URL of the on-demand HLS transcode
// tier. Each play session carries a fresh token so concurrent viewers keep
// separate segment caches (ADR-002 多终端并发); audio selects which audio track
// is transcoded (default 0).
export function hlsPlaylistUrl(id: string, session: string, audio = 0): string {
  const params = new URLSearchParams({ session })
  if (audio > 0) params.set('audio', String(audio))
  return `/api/stream/${id}/hls/playlist.m3u8?${params.toString()}`
}

export function coverUrl(id: string, thumb = false): string {
  return `/api/stream/${id}/cover${thumb ? '?thumb=1' : ''}`
}

export function subtitleUrl(id: string, track?: number): string {
  return `/api/stream/${id}/subtitle${track !== undefined ? `?track=${track}` : ''}`
}

// SubtitleTrack is one subtitle source of a video (a sidecar file next to the
// video, an embedded text stream of the container, or a bitmap track).
// playable tracks can feed the player's <track> menu; a bitmap track (PGS/
// VobSub, playable=false) cannot be converted to text and only works burned
// into the picture by the format factory.
export interface SubtitleTrack {
  kind: 'sidecar' | 'embedded'
  index?: number
  codec?: string
  label: string
  playable?: boolean
}

export function fetchVideoSubtitles(id: string): Promise<{ subtitles: SubtitleTrack[] }> {
  return api<{ subtitles: SubtitleTrack[] }>(`/api/videos/${id}/subtitles`)
}

// AudioTrack is one audio track of a video the player can switch to (multi-track
// containers like 国语/粤语 MKVs). Index is the 0-based audio-stream ordinal,
// passed to the stream endpoints as ?audio=.
export interface AudioTrack {
  index: number
  codec?: string
  language?: string
  label: string
  channels?: number
}

export function fetchVideoAudioTracks(id: string): Promise<{ audio: AudioTrack[] }> {
  return api<{ audio: AudioTrack[] }>(`/api/videos/${id}/audio`)
}
